app/ciim/client.py

```diff
@@ -0,0 +1,174 @@
+import json
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, Any, Dict, Iterable, Optional, Tuple, Type
+
+import requests
+from app.records.models import Record
+from django.utils.functional import cached_property
+
+from .exceptions import (
+    ClientAPIBadRequestError,
+    ClientAPICommunicationError,
+    ClientAPIInternalServerError,
+    ClientAPIServiceUnavailableError,
+    DoesNotExist,
+    MultipleObjectsReturned,
+)
+
+if TYPE_CHECKING:
+    from app.ciim.models import APIModel
+
+
+class ResultList:
+    """
+    A convenience class that lazily converts a raw list of "hits" (from various
+    API endpoints) into instances of `item_type` when iterated.
+    """
+
+    def __init__(
+        self,
+        hits: Sequence[Dict[str, Any]],
+        item_type: Type,
+    ):
+        self._hits = hits or []
+        self.item_type = item_type
+
+    @cached_property
+    def hits(self) -> Tuple["APIModel"]:
+        """
+        Return a tuple of APIModel instances representative of the raw `_hits`
+        data. The return value is cached to support reuse without any
+        transformation overhead.
+        """
+        return tuple(
+            self.item_type.from_api_response(h) if isinstance(h, dict) else h
+            for h in self._hits
+        )
+
+    def __iter__(self) -> Iterable["APIModel"]:
+        yield from self.hits
+
+    def __len__(self) -> int:
+        return len(self._hits)
+
+    def __bool__(self) -> bool:
+        return bool(self._hits)
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__} {self.hits}>"
+
+
+class ClientAPI:
```

Comment:
`ClientAPI` isn't very descriptive. Would this be better as something like `RosettaAPI` seeing as that is what we are querying?

Reply:
It used to be `KongClient` before `ClientAPI`. Wanted to use a more generic name as these names Kong, Rosetta can change in the future. The wording Client in the name says that its a typically a requesting unit which interacts with an external API. It acts as an intermediary so that the application can send requests to the API and get a response.

The environment variable `CLIENT_BASE_URL` that uses  Rosetta API can be changed to `ROSETTA_CLIENT_BASE_URL` to make it more clear and what  `settings.CLIENT_BASE_URL` is using. 

I've updated the doc comments for `ClientAPI` to be more clearer.

Reply:
But there was a huge difference between the Kong and Rosetta APIs.

I would have personally preferred that we have a `KongAPI` class that stayed the same while we worked on a new `RosettaAPI` class. That way we could have more easily switched between the two.

Having a single API class that we change when we switch to a new API is rather limiting.

Reply:
> That way we could have more easily switched between the two.
This is not a PR for switching.